### PR TITLE
#569:Partial Fix: Fix for Users Blocking

### PR DESCRIPTION
**Pull request by varadarajana**
Please find the 2 files containing the partial fixes for blocking users. I will create one more for organization.

### CODE REVIEW

File: github/orgs_members_test.go [hunk 1]
```diff
@@ -395,7 +395,11 @@ func TestOrganizationsService_ListPendingOrgInvitations(t *testing.T) {
 	})
 
 	opt := &ListOptions{Page: 1}
+<<<<<<< HEAD
 	invitations, _, err := client.Organizations.ListPendingOrgInvitations(context.Background(), 1, opt)
+=======
+	invitations, _, err := client.Organizations.ListPendingOrgInvitations(1, opt)
+>>>>>>> 1bd4328a520dfc6d58f92e0f441f3f56718b725d
```
Comment (shurcooL) [inline github/orgs_members_test.go]: This unresolved merge conflict shouldn't be here.

File: github/users_blocking.go [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+// license that can be found in the LICENSE file.
+
+package github
+
+import (
+	"context"
+	"fmt"
+)
+
+// ListBlockedUsers lists all the blocked users by the authenticated user
```
Comment (shurcooL) [inline github/users_blocking.go]: Sentences end with a period.

File: github/users_blocking.go [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+		return nil, resp, err
+	}
+
+	return blockedUsers, resp, nil
+}
+
+// CheckIfUserIsBlocked allows the authenticated user to check if the other user is blocked
+//
+// GitHub API docs: https://developer.github.com/v3/users/blocking/#check-whether-youve-blocked-a-user
+func (s *UsersService) CheckIfUserIsBlocked(ctx context.Context, user string) (bool, *Response, error) {
```
Comment (shurcooL) [inline github/users_blocking.go]: I'm thinking we can shorten/simplify the method name to something like `IsUserBlocked`. @gmlewis, what do you think about the naming here?

File: github/users_blocking.go [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+	u, err := addOptions(u, opt)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	req, err := s.client.NewRequest("GET", u, nil)
+	if err != nil {
+		return nil, nil, err
+	}
+
```
Comment (shurcooL) [inline github/users_blocking.go]: According to the the GitHub API docs:

image

I think you're missing that preview header.

File: github/users_blocking.go (hunk 3 above)
Comment (shurcooL) [inline github/users_blocking.go]: Some precedent:

https://godoc.org/github.com/google/go-github/github#UsersService.IsFollowing

File: github/users_blocking.go (hunk 4 above)
Comment (varadarajana) [inline github/users_blocking.go]: @shurcooL Thank you, I missed the preview header, now it is added. Can you please review?

File: github/orgs_members_test.go (hunk 1 above)
Comment (varadarajana) [inline github/orgs_members_test.go]: @shurcooL this was part of messed up fork of mine. I have corrected it and now the latest check in has the right changes.

File: github/users_blocking.go [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+		return nil, resp, err
+	}
+
+	return blockedUsers, resp, nil
+}
+
+// CheckIfUserIsBlocked allows the authenticated user to check if the other user is blocked.
+//
+// GitHub API docs: https://developer.github.com/v3/users/blocking/#check-whether-youve-blocked-a-user
+func (s *UsersService) CheckIfUserIsBlocked(ctx context.Context, user string) (bool, *Response, error) {
```
Comment (shurcooL) [inline github/users_blocking.go]: @varadarajana, did you see my comment at https://github.com/google/go-github/pull/600#discussion_r108045741?

Comment (varadarajana) [inline github/users_blocking.go]: @shurcooL I will change the method name and submit, I was waiting since you had asked this question to @gmlewis

Comment (shurcooL) [inline github/users_blocking.go]: Fair enough. Let's go with `IsBlocked` for now, I feel pretty good about it. If there are additional concerns or improvement suggestions, we can still change it.

Comment (varadarajana) [inline github/users_blocking.go]: @shurcooL I have corrected it to IsBlocked.

**Comment by shurcooL** [comment]
Not seeing any high level issues. Just need to review it at a lower level.

File: github/users_blocking.go [hunk 6]
```diff
@@ -0,0 +1,91 @@
+// Copyright 2013 The go-github AUTHORS. All rights reserved.
```
Comment (gmlewis) [inline github/users_blocking.go]: 2017

File: github/users_blocking.go [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+	}
+
+	return blockedUsers, resp, nil
+}
+
+// IsBlocked allows the authenticated user to check if the other user is blocked.
+//
+// GitHub API docs: https://developer.github.com/v3/users/blocking/#check-whether-youve-blocked-a-user
+func (s *UsersService) IsBlocked(ctx context.Context, user string) (bool, *Response, error) {
+	u := fmt.Sprintf("users/blocks/%v", user)
```
Comment (gmlewis) [inline github/users_blocking.go]: According to the docs, this URL is: `"user/blocks/%v"`

File: github/users_blocking.go [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+	resp, err := s.client.Do(ctx, req, nil)
+	isBlocked, err := parseBoolResponse(err)
+	return isBlocked, resp, err
+}
+
+// BlockUser allows authenticated User to block another User.
+//
+// GitHub API docs: https://developer.github.com/v3/users/blocking/#block-a-user
+func (s *UsersService) BlockUser(ctx context.Context, user string) (*Response, error) {
+	u := fmt.Sprintf("users/blocks/%v", user)
```
Comment (gmlewis) [inline github/users_blocking.go]: `"user/blocks/%v"`

File: github/users_blocking.go [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+	req.Header.Set("Accept", mediaTypeBlockUsersPreview)
+
+	return s.client.Do(ctx, req, nil)
+}
+
+// UnblockUser allows authenticated User to unblock BlockedUser.
+//
+// GitHub API docs: https://developer.github.com/v3/users/blocking/#unblock-a-user
+func (s *UsersService) UnblockUser(ctx context.Context, user string) (*Response, error) {
+	u := fmt.Sprintf("users/blocks/%v", user)
```
Comment (gmlewis) [inline github/users_blocking.go]: `"user/blocks/%v"`

File: github/users_blocking_test.go [hunk 10]
```diff
@@ -0,0 +1,90 @@
+// Copyright 2014 The go-github AUTHORS. All rights reserved.
```
Comment (gmlewis) [inline github/users_blocking_test.go]: 2017

File: github/users_blocking_test.go [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+	if want := false; isBlocked != want {
+		t.Errorf("Users.IsBlocked returned %+v, want %+v", isBlocked, want)
+	}
+}
+
+func TestUsersService_BlockUser(t *testing.T) {
+	setup()
+	defer teardown()
+
+	mux.HandleFunc("/users/blocks/u", func(w http.ResponseWriter, r *http.Request) {
```
Comment (gmlewis) [inline github/users_blocking_test.go]: `"user/blocks/u"`

File: github/users_blocking_test.go [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+	if err != nil {
+		t.Errorf("Users.BlockUser returned error: %v", err)
+	}
+}
+
+func TestUsersService_UnblockUser(t *testing.T) {
+	setup()
+	defer teardown()
+
+	mux.HandleFunc("/users/blocks/u", func(w http.ResponseWriter, r *http.Request) {
```
Comment (gmlewis) [inline github/users_blocking_test.go]: `"user/blocks/u"`

File: github/users_blocking.go (hunk 6 above)
Comment (varadarajana) [inline github/users_blocking.go]: @gmlewis this is fixed.

File: github/users_blocking.go (hunk 7 above)
Comment (varadarajana) [inline github/users_blocking.go]: @gmlewis This is fixed.

File: github/users_blocking.go (hunk 8 above)
Comment (varadarajana) [inline github/users_blocking.go]: @gmlewis this is fixed

File: github/users_blocking.go (hunk 9 above)
Comment (varadarajana) [inline github/users_blocking.go]: @gmlewis this is fixed

File: github/users_blocking_test.go (hunk 10 above)
Comment (varadarajana) [inline github/users_blocking_test.go]: @gmlewis this is fixed

File: github/users_blocking_test.go (hunk 11 above)
Comment (varadarajana) [inline github/users_blocking_test.go]: @gmlewis this is fixed

File: github/users_blocking_test.go (hunk 12 above)
Comment (varadarajana) [inline github/users_blocking_test.go]: @gmlewis this is fixed

**Comment by varadarajana** [comment]
@gmlewis can you please review these changes?

File: github/users_blocking.go [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+	var blockedUsers []*User
+	resp, err := s.client.Do(ctx, req, &blockedUsers)
+	if err != nil {
+		return nil, resp, err
+	}
+
+	return blockedUsers, resp, nil
+}
+
+// IsBlocked allows the authenticated user to check if the other user is blocked.
```
Comment (shurcooL) [inline github/users_blocking.go]: Instead of saying "MethodName allows the authenticated user to do something", we can be more direct and say "MethodName does something for authenticated user."

I would rephrase this as:

```
// IsBlocked reports whether specified user is blocked for the authenticated user.
```

File: github/users_blocking.go [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+
+	// TODO: remove custom Accept header when this API fully launches.
+	req.Header.Set("Accept", mediaTypeBlockUsersPreview)
+
+	resp, err := s.client.Do(ctx, req, nil)
+	isBlocked, err := parseBoolResponse(err)
+	return isBlocked, resp, err
+}
+
+// BlockUser allows authenticated User to block another User.
```
Comment (shurcooL) [inline github/users_blocking.go]:
```
// BlockUser blocks specified user for the authenticated user.
```

File: github/users_blocking.go [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+		return nil, err
+	}
+
+	// TODO: remove custom Accept header when this API fully launches.
+	req.Header.Set("Accept", mediaTypeBlockUsersPreview)
+
+	return s.client.Do(ctx, req, nil)
+}
+
+// UnblockUser allows authenticated User to unblock BlockedUser.
```
Comment (shurcooL) [inline github/users_blocking.go]:
```
// UnblockUser unblocks specified user for the authenticated user.
```

**Comment by shurcooL** [comment]
Thanks @varadarajana!
